### PR TITLE
Create a simple example of how to use zookeeper with finagle

### PR DESCRIPTION
**Pull request by sunnykaka**
Problem

Lack of examples and documents about how to using zookeeper with finagle.

Solution

Create a simple example of how to use zookeeper with finagle.

Result

It's a example, it does not affect other modules.

### CODE REVIEW

File: finagle-example/src/main/scala/com/twitter/finagle/example/zookeeper/ServiceConsumer.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  *
+  * Run ServiceProvider before ServiceConsumer
+  *
+  */
+object ServiceConsumer {
+
+  def main(args: Array[String]): Unit = {
+
+    //use zookeeper to discover service
+    val client: Service[Request, Response] = ClientBuilder()
```
Comment (mosesn) [inline finagle-example/src/main/scala/com/twitter/finagle/example/zookeeper/ServiceConsumer.scala]: would you mind using a finagle 6-style client?  This instead becomes:

```scala
Http
  .client
  .withSessionPool.maxSize(10)
  .newService(ServiceProvider.buildConsumerPath(EchoServer.servicePath), "http-client")
```

The new version of finagle that supports the with-style parameters should be released later this week.

File: finagle-example/src/main/scala/com/twitter/finagle/example/zookeeper/ServiceProvider.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  object EchoServer {
+    val servicePath = "/services/echo"
+  }
+
+  def main(args: Array[String]): Unit = {
+
+    //start your service on port 8080
+    val server: Server = ServerBuilder()
```
Comment (mosesn) [inline finagle-example/src/main/scala/com/twitter/finagle/example/zookeeper/ServiceProvider.scala]: Ditto here:

```scala
Http.server.withLabel("echoserver").serve(":8080", new EchoServer)
```

**Comment by mosesn** [comment]
LGTM, but it would be nice to use zk2 instead of zk, which is what we use internally, and we think it's more reliable.  I'll get some folks to come take a look at your usage :+1:

**Comment by chchen** [comment]
zk2 only works for the resolution path, right? For announcing, zk is the only supported scheme I think.

**Comment by sunnykaka** [comment]
Thanks @mosesn , I will use the finagle 6-style api.

**Comment by sunnykaka** [comment]
Hello @chchen , I didn't use zk2 in my project yet.
I'll try it.

File: finagle-example/src/main/scala/com/twitter/finagle/example/zookeeper/ServiceConsumer.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  * Run ServiceProvider before ServiceConsumer
+  *
+  */
+object ServiceConsumer {
+
+  def main(args: Array[String]): Unit = {
+
+    //use zookeeper to discover service
+    val client = Http.client.
+      withSessionPool.maxSize(10).
```
Comment (mosesn) [inline finagle-example/src/main/scala/com/twitter/finagle/example/zookeeper/ServiceConsumer.scala]: can we move the . to be in front?  that's the typical style we use inf inagle

File: finagle-example/src/main/scala/com/twitter/finagle/example/zookeeper/ServiceProvider.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
+package com.twitter.finagle.example.zookeeper
+
+import java.net.InetSocketAddress
```
Comment (mosesn) [inline finagle-example/src/main/scala/com/twitter/finagle/example/zookeeper/ServiceProvider.scala]: do you mind sorting your imports?

File: finagle-example/src/main/scala/com/twitter/finagle/example/zookeeper/ServiceProvider.scala [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    //syntax:          schema!host!path!shardId
+    //schema:          for server, use zk, for client, use zk2
+    //host:            zookeeper connection string
+    //path:            service registration path in zookeeper
+    //shardId:         it's used internally by Twitter, can be set to 0 in most cases
+    val echoServicePath = buildProviderPath(EchoServer.servicePath)
+
+    //start your service on port 8080 and register to zookeeper
+    Await.ready(
+      Http.server.
```
Comment (mosesn) [inline finagle-example/src/main/scala/com/twitter/finagle/example/zookeeper/ServiceProvider.scala]: dots in front :+1:

**Comment by mosesn** [comment]
LGTM except for a couple nits.  Sorry for the delay, github doesn't tell us when you merge in a new commit.

**Comment by sunnykaka** [comment]
Hi @mosesn , I have fix it.

**Comment by mosesn** [comment]
LGTM, thanks!

**Comment by mosesn** [comment]
@sunnykaka this was merged internally, thanks for the contribution!  :fire:

**Comment by mosesn** [comment]
This made it to develop here: https://github.com/twitter/finagle/commit/766f38974663b80509f7d19ae9263c659b11d8a6.  Really appreciate the PR!  :fire:

**Pull request closed without merging**